### PR TITLE
Insert pipe character when editing multiline strings

### PR DESCRIPTION
**Pull request by Karola766**
Partially solves #859. Doesn't work for copypasted strings yet, though.

The behaviour isn't identical to that demonstrated in gifs under issue #859. For now, before making a TextEdit, Provider checks whether there was a pipe in the last line, even, if it's the first line of the string. It doesn't yet add .stripMargin after the end of the string, either. So we might want to expand on that.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    val bit = toInput.text.substring(0, pos.start)
+    val lastNewline = bit.lastIndexOf("\n")
+    val lastPipe = bit.lastIndexOf("|")
+    val length = toInput.text
+      .substring(lastNewline, lastPipe)
+      .length - 1
+    space * length
+  }
+
+  private def longString(text: String, token: Token): Boolean = {
```
Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: those private defs should be below the one method uses them (or even defined within this method)

Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: `isLongString` sound a bit nicer nicer when read inside an if: 
`if((...) and isLongString(foo)`

Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: maybe `isMultilineString` ? one-line string can also be very long ;)

File: metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    text.substring(token.start, token.start + 3).equals(quot)
+  }
+
+  private def inToken(pos: meta.Position, token: Token): Boolean = {
+    pos.start >= token.start && pos.end <= token.end
+  }
+
+  private def pipeInScope(pos: meta.Position, text: String): Boolean = {
+    val lastNew =
```
Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: how about `lastNewline` ?

File: metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      params: DocumentOnTypeFormattingParams
+  ): Future[Option[TextEdit]] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+    if (source.exists) {
+      val range = new Range(params.getPosition, params.getPosition)
+      val pos = params.getPosition.toMeta(source.toInput)
+      val sourceText = source.toInput.text
+      val tokens = source.toInput.tokenize.toOption
+      var edit: Option[TextEdit] = None
+      tokens.get.foreach {
```
Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: why `get` ? what if there are no tokens? maybe `getOrElse(Nil)` would be better?
or something like this?
```scala
if(pipeInScope(pos, sourceText){
  val tokenized = source.toInput.tokenize.toOption
  tokenized.flatMap{ tokens => 
    tokens.collectFirst{ 
      case token: Constant.String if inToken(pos, token) && longString(sourceText, token) => 
        new TextEdit(range, indent(source.toInput, pos) + "|")
      }
  }
} else {
  None
}
```

File: metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+          if (inToken(pos, token) && longString(sourceText, token) && pipeInScope(
+              pos,
+              sourceText
+            ))
+            edit = Some(new TextEdit(range, indent(source.toInput, pos) + "|"))
+        case _ => None
+      }
+      Future.successful(edit)
+    } else {
+      Future.failed(new Exception("There is no source"))
```
Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: Why failure instead of a `None` ?

**Comment by tgodzik** [comment]
Checking this in VSCode - it seems doesn't yet work. We need to change some things.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 5]
```diff
@@ -841,6 +849,16 @@ class MetalsLanguageServer(
       )
     }
 
+  @JsonRequest("textDocument/onTypeFormatting")
```
Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: does it work in vscode?  The [specification](https://microsoft.github.io/language-server-protocol/specification#textDocument_onTypeFormatting) expects a `List`, not an `Option[TextEdit]`.

File: metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    text
+      .substring(lastNewline + 1, pos.start)
+      .contains("|")
+  }
+
+  def format(
+      params: DocumentOnTypeFormattingParams
+  ): Future[Option[TextEdit]] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+    var edit: Option[TextEdit] = None
```
Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: we can remove this `var` if we add some else blocks:
```scala
else Future.successful(None)
```

File: metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        Input.VirtualFile(source.toString(), sourceText)
+      )
+      if (pipeInScope(pos, sourceText)) {
+        val tokens =
+          Input.VirtualFile(source.toString(), sourceText).tokenize.toOption
+        tokens.flatMap { tokens =>
+          tokens.collectFirst {
+            case token: Constant.String
+                if inToken(pos, token) && isLongString(sourceText, token) =>
+              edit = Some(new TextEdit(range, indent(sourceText, pos) + "|"))
```
Comment (marek1840) [inline metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: Let's just return the `new TextEdit(...)` here. as a result, the `tokens.flatMap` on line 62 will be returning this edit.
```scala
if (pipeInScope(pos, sourceText)) {
  val tokens = Input.VirtualFile(...)
  tokens.flatMap { tokens => tokens.collectFirst{ case (...) if (...) => new TextEdit(...) } }
} else Future.successfull(None)
```
or we could do:
```scala
tokens match {
  case None => 
    Future.successful(None)
  case tokens => 
    tokens.collectFirst{ case (...) if (...) => new TextEdit(...) }
}
```

whatever you find more readable

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala (hunk 5 above)
Comment (Karola766) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: it doesn't, and you're right, I haven't noticed. But I've changed my ontypeformatting product to a list and the server still doesn't seem to be responding properly, so it's probably something deeper, too.

File: metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala (hunk 7 above)
Comment (Karola766) [reply metals/src/main/scala/scala/meta/internal/metals/OnTypeFormattingProvider.scala]: ? how can we return Future.succesful(None) when we want to return Future[TextEdit]? wouldn't a None be more fitting for Option[TextEdit]?

**Pull request merged**